common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -436,17 +456,16 @@ public StreamCallbackWithID receiveBlockDataAsStream(PushBlockStream msg) {
     // getting killed. When this happens, we need to distinguish the duplicate blocks as they
     // arrive. More details on this is explained in later comments.
 
-    // Track if the block is received after shuffle merge finalized or from an older
-    // shuffleMergeId attempt.
-    final boolean isStaleBlockOrTooLate = partitionInfoBeforeCheck == null;
     // Check if the given block is already merged by checking the bitmap against the given map
     // index
-    final AppShufflePartitionInfo partitionInfo = isStaleBlockOrTooLate ? null :
+    final AppShufflePartitionInfo partitionInfo = isStaleBlock || isTooLate ? null :
       partitionInfoBeforeCheck.mapTracker.contains(msg.mapIndex) ? null : partitionInfoBeforeCheck;
     if (partitionInfo != null) {
       return new PushBlockStreamCallback(
         this, appShuffleInfo, streamId, partitionInfo, msg.mapIndex);
     } else {
+      final boolean finalIsTooLate = isTooLate;
+      final boolean finalIsStaleBlock = isStaleBlock;
```

Comment:
```suggestion
      final BlockPushNonFatalFailure exception = isTooLate ? createTooLate .. : (isStaleBlock ? createStale ..  : null);
```

and `if (null != exception) throw exception` in onComplete handler.